fixes `./...` syntax in `tinygo test`

jcchavezs opened the pull request:
Adds proper support for `./...` syntax which was broken or maybe never intended to be supported and it only somehow worked because we use `go list` underneath. Allowing to use `./...` is actually a good experience in `go test`.

For example, running the tinygo@0.23.0 like:

```
$ tinygo test ./tests/testing/threedots/...
ok      github.com/tinygo-org/tinygo/tests/testing/threedots/pass       0.106s
```

whereas `./tests/testing/threedots` has two packages.

Closes #2892

dkegel-fastly commented:
Could you link to (and if needed, file) a tinygo issue describing the problem you're trying to solve?  Thanks!

dkegel-fastly commented:
Could you link to (and if needed, file) a tinygo issue describing the problem you're trying to solve?  Thanks!

jcchavezs commented on loader/loader.go:
```diff
@@ -89,7 +90,7 @@ type Package struct {
 // package). Call .Parse() afterwards to parse all Go files (including CGo
 // processing, if necessary).
-func Load(config *compileopts.Config, inputPkgs []string, clangHeaders string, typeChecker types.Config) (*Program, error) {
+func Load(config *compileopts.Config, inputPkg string, clangHeaders string, typeChecker types.Config) (*Program, error) {
```
I turned this into a single package for a few reasons:
1. It is used for single package only everywhere.
2. It is easier to deal with a single package.
3. [This error](https://github.com/tinygo-org/tinygo/pull/2885/files#diff-8e9af1e483632eac17610237caaa5f879ea7bfb3e1e1ea228e45a17c770b9432L231) is only looking at one package and it is actually thought of a single package

jcchavezs commented:
@dkegel-fastly I linked the problem in the description. Do you want me to create a github issue?

dkegel-fastly commented:
Yeah, at least for now, it would help if you had a github issue / github PR pair for each individual problem you are solving.  (I used to try to do multiple problems in one PR, but life is better with just one problem per PR, linked to a clear issue with reproducer.)

dkegel-fastly committed:
smoketest: add regression test for 'tinygo test ./...', see #2892

tests/testing/recurse has two directories with tests;
"make smoketest" now does "tinygo test ./..." in that directory
and fails if it does not run both directories' tests.

jcchavezs commented:
Did a cherry-pick of [this commit](https://github.com/tinygo-org/tinygo/pull/2894/commits/8bdb2a1d6d70871172f2dcb889a11bd62d0cda42) to try the smoke test cc @dkegel-fastly

jcchavezs commented:
It is green :D @dkegel-fastly

dkegel-fastly commented:
I would suggest cleaning up the commit history in this branch a bit... make it looks like you got it right the first try.  No need to show dirty laundry!

jcchavezs commented:
I can do that but I wonder if we keep this as it is because there are some
learnings in the mistakes and do the clean up of messaging by squashing the
merge?

deadprogram commented:
Hello, @jcchavezs and @dkegel-fastly 

Having atomic commits and a clean history is part of our merge/rebase criteria. If you want to preserve additional info, I would suggest keep that in a branch on a fork.

Hopefully that helps clarify. Thank you!

jcchavezs commented:
Thanks for the clarification. I will do so.

jcchavezs committed:
fix: fixes tinygo test ./... syntax.

jcchavezs commented:
I cleaned up the commits.

jcchavezs commented:
Can we get this merged please?

dkegel-fastly commented:
Give it a couple days :-)

jcchavezs commented:
Shall we?

dgryski commented:
I'll review this today.

dgryski reviewed (approved):
LGTM

deadprogram commented:
Thank you very much for working on this @jcchavezs now merging.
